### PR TITLE
refactor: implement RouteGenerator for backend routes

### PR DESCRIPTION
**Pull request by nareddyt**
`BackendGenerator` is a `RouteGenerator` to configure routes to the local or remote backend service. It configures all aspects of the routes that are forward to a backend cluster, including sorting them via HTTP pattern trie for correctness. Please see `backend_test.go` to double check the generated routes.

Fork a few methods from `ServiceInfo` into utility file. Refactor methods to return `map[selector]X` as pre-processed input for the generator.

TODO in follow-up CL:
- CORS pass-through routes to backends (when `AllowCors` is enabled).
- Set route deadlines and timeouts.

NOT used in prod yet

### CODE REVIEW

File: src/go/configgenerator/routegen/backend.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+		backendCluster := g.BackendClusterBySelector[selector]
+		if backendCluster == nil {
+			return nil, fmt.Errorf("could not find any backend cluster for selector %q", selector)
+		}
+
+		methodCfg := &helpers.MethodCfg{
+			OperationName:      selector,
+			BackendClusterName: backendCluster.Name,
+			HostRewrite:        backendCluster.HostName,
+			// TODO(nareddyt)
```
Comment (shuoyang2016) [inline src/go/configgenerator/routegen/backend.go:64]: What's this TODO?

File: src/go/configgenerator/routegen/backend.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+// BackendGenerator is a RouteGenerator to configure routes to the local
+// or remote backend service.
+type BackendGenerator struct {
+	HTTPPatterns             *httppattern.MethodSlice
+	BackendClusterBySelector map[string]*BackendClusterSpecifier
+	BackendRouteGen          *helpers.BackendRouteGenerator
+}
+
+// NewBackendRouteGensFromOPConfig creates BackendGenerator
+// from OP service config + descriptor + ESPv2 options.
```
Comment (shuoyang2016) [inline src/go/configgenerator/routegen/backend.go:23]: nit: is descriptor used for route generator?

**Review by TAOXUY — APPROVED** [review]

File: src/go/configgenerator/routegen/backend.go (hunk 2 above)
Comment (nareddyt) [reply src/go/configgenerator/routegen/backend.go]: No. I guess service config contains descriptor internally, but i'll remove it from the comment.

File: src/go/configgenerator/routegen/util.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+			method: rule.GetCustom().GetKind(),
+		}, nil
+	default:
+		return nil, fmt.Errorf("error parsing http rule type: unsupported http method %T", rule.GetPattern())
+	}
+}
+
+// GetHTTPRuleBySelectorFromOPConfig pre-processes the service config to return
+// a map of selector to the corresponding HTTP rule.
+func GetHTTPRuleBySelectorFromOPConfig(serviceConfig *servicepb.Service, opts options.ConfigGeneratorOptions) map[string]*annotationspb.HttpRule {
```
[the remark this answers is not in the document]
Comment (nareddyt) [reply src/go/configgenerator/routegen/util.go]: Done

File: src/go/configgenerator/routegen/util.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+}
+
+// ParseHTTPPatternsBySelectorFromOPConfig parses the service config into a list
+// of internal HTTP pattern representations, keyed by OP selector.
+//
+// Note: HTTP here refers to pattern serving as a route on the HTTP protocol.
+// It does not differentiation between HTTP vs gRPC backend.
+// By default, this function will generate HTTP patterns for gRPC backends.
+//
+// Forked from `service_info.go: processHttpRule()`
```
[the remark this answers is not in the document]
Comment (nareddyt) [reply src/go/configgenerator/routegen/util.go:114]: Done

File: src/go/configgenerator/routegen/utils_test.go [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+	servicepb "google.golang.org/genproto/googleapis/api/serviceconfig"
+	apipb "google.golang.org/genproto/protobuf/api"
+)
+
+func TestParseSelectorsFromOPConfig(t *testing.T) {
+	testdata := []struct {
+		name          string
+		serviceConfig *servicepb.Service
+		opts          options.ConfigGeneratorOptions
+		want          []string
```
[the remark this answers is not in the document]
Comment (nareddyt) [reply src/go/configgenerator/routegen/utils_test.go:20]: No need to make test vars longer, as the test name already indicates what is being tested.

[go/gotip/episodes/21](https://goto.google.com/gotip/episodes/21) uses `want` as well

File: src/go/configgenerator/routegen/util.go [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+	address := fmt.Sprintf("%v:%v", hostname, port)
+	return &BackendClusterSpecifier{
+		Name:     clustergen.RemoteAddressToClusterName(address),
+		HostName: hostname,
+	}, nil
+}
+
+// GetBackendRuleBySelectorFromOPConfig pre-processes the service config to
+// return a map of selector to the corresponding backend rule.
+func GetBackendRuleBySelectorFromOPConfig(serviceConfig *servicepb.Service, opts options.ConfigGeneratorOptions) map[string]*servicepb.BackendRule {
```
[the remark this answers is not in the document]
Comment (nareddyt) [reply src/go/configgenerator/routegen/util.go]: Changed to `Precompute`

File: src/go/configgenerator/routegen/backend.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+		routes = append(routes, methodRoutes...)
+	}
+
+	return routes, nil
+}
+
+// sortHttpPatterns implements go/esp-v2-route-match-ordering-implementation.
+// Sorting is needed for route match correctness.
+//
+// Forked from `route_generator.go: sortHttpPatterns()`
```
[the remark this answers is not in the document]
Comment (nareddyt) [reply src/go/configgenerator/routegen/backend.go:81]: Yes, I have b/288170480 to remove all of `ServiceInfo`. Will take a few more CLs.

File: src/go/configgenerator/routegen/backend.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nareddyt) [reply src/go/configgenerator/routegen/backend.go]: Added more context inline

**Review by shuoyang2016 — APPROVED** [review]

**Pull request merged**